dlc/src/main/scala/org/bitcoins/dlc/DLCMessage.scala

```diff
@@ -2,28 +2,234 @@ package org.bitcoins.dlc
+      val rValue = ECPublicKey(bytes.drop(33))
+
+      // TODO: validate rValue once the libsecp functionality is added
+
+      OracleInfo(pubkey, rValue)
+    }
+  }
+
+  case class ContractInfo(contractInfo: Map[Sha256DigestBE, Satoshis])
+      extends NetworkElement {
+    override def bytes: ByteVector = {
+      contractInfo.foldLeft(ByteVector.empty) {
+        case (vec, (digest, sats)) => vec ++ digest.bytes ++ sats.bytes
+      }
+    }
+  }
+
+  object ContractInfo extends Factory[ContractInfo] {
+
+    override def fromBytes(bytes: ByteVector): ContractInfo = {
+      @tailrec
+      def loop(
+          remainingBytes: ByteVector,
+          accum: Seq[(Sha256DigestBE, Satoshis)]): Seq[
+        (Sha256DigestBE, Satoshis)] = {
+        if (remainingBytes.size < 40) {
+          accum
+        } else {
+          val relevantBytes = remainingBytes.take(40)
+          val digest = Sha256DigestBE(relevantBytes.take(32))
+          val sats = Satoshis(relevantBytes.takeRight(8))
+          loop(remainingBytes.drop(40), accum :+ (digest, sats))
+        }
+      }
+      ContractInfo(loop(bytes, Nil).toMap)
+    }
+  }
 
   case class DLCOffer(
-      contractInfo: Map[Sha256DigestBE, Satoshis],
+      contractInfo: ContractInfo,
       oracleInfo: OracleInfo,
       extPubKey: ExtPublicKey,
       totalCollateral: Satoshis,
       fundingInputs: Vector[(TransactionOutPoint, TransactionOutput)],
       changeAddress: Bech32Address,
       feeRate: SatoshisPerVirtualByte,
       timeouts: DLCTimeouts)
-      extends DLCMessage
+      extends DLCMessage {
+
+    def toJson: Value = {
+      val contractInfosJson =
+        contractInfo.contractInfo
+          .map(
+            info =>
+              mutable.LinkedHashMap("sha256" -> Str(info._1.hex),
+                                    "sats" -> Num(info._2.toLong)))
+
+      val fundingInputsJson =
+        fundingInputs
+          .map(
+            input =>
+              mutable.LinkedHashMap("outpoint" -> Str(input._1.hex),
+                                    "output" -> Str(input._2.hex)))
+
+      val timeoutsJson =
+        mutable.LinkedHashMap(
+          "penalty" -> Num(timeouts.penaltyTimeout),
+          "contractMaturity" -> Num(timeouts.contractMaturity.toUInt32.toLong),
+          "contractTimeout" -> Num(timeouts.contractTimeout.toUInt32.toLong)
+        )
+
+      Obj(
+        mutable.LinkedHashMap[String, Value](
+          "contractInfo" -> contractInfosJson,
+          "oracleInfo" -> Str(oracleInfo.hex),
+          "extPubKey" -> Str(extPubKey.hex),
+          "totalCollateral" -> Num(totalCollateral.toLong),
+          "fundingInputs" -> fundingInputsJson,
+          "changeAddress" -> Str(changeAddress.value),
+          "feeRate" -> Num(feeRate.toLong),
+          "timeouts" -> timeoutsJson
+        )
+      )
+    }
+
+    def toJsonStr: String = toJson.toString()
+  }
+
+  object DLCOffer {
+
+    def fromJson(js: Value): DLCOffer = {
```

Comment:
This all seems much grosser than using play Json framework :( I'm fine with this for now but we should aim to change this to the nicer easier to deal with stuff later (file an issue maybe?)

Reply:
I don't believe the `cli` project can use play Json :(

Reply:
@Christewart

Reply:
I believe you can still get around it by having `toJsonStr` and `fromJsonStr` implemented with a play json backend and then just use the standard ujson <-> String translation

Reply:
I don't think macros are supported by native image.